cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py

```diff
@@ -103,6 +103,22 @@ def upgrade_charm():
     if hookenv.config().get('ingress'):
         kubectl_success('delete', 'rc', 'nginx-ingress-controller')
 
+        # these moved into a different namespace for 1.12
+        kubectl_success('delete', 'rc', 'default-http-backend')
+        kubectl_success('delete', 'svc', 'default-http-backend')
+        kubectl_success('delete', 'serviceaccount',
+                        'nginx-ingress-{}-serviceaccount'.format(
+                            hookenv.service_name()))
+        kubectl_success('delete', 'clusterrolebinding',
+                        'nginx-ingress-clusterrole-nisa-{}-binding'.format(
+                            hookenv.service_name()))
+        kubectl_success('delete', 'configmap',
+                        'nginx-load-balancer-{}-conf'.format(
+                            hookenv.service_name()))
+        kubectl_success('delete', 'ds',
+                        'nginx-ingress-{}-controller'.format(
+                            hookenv.service_name()))
```

Comment:
This could get a bit messy if any of these deletes fail (e.g. apiserver briefly goes down as kubernetes-master is upgraded at the same time). Seems like we could fail to delete the old controller, which would clash with the new one as they both try to claim port 443 on the host.

I recommend adding retry logic to this. Might be best to extract this out to a handler that can be re-run until everything is cleaned up.